pkg/tfbridge/tests/provider_test.go

```diff
@@ -129,6 +130,363 @@ func TestInputsConfigModeEqual(t *testing.T) {
+	t.Parallel()
+	crosstests.Create(t,
+		map[string]*schema.Schema{
+			"f0": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"x": {Optional: true, Type: schema.TypeString},
+					},
+				},
+			},
+		},
+		cty.ObjectVal(map[string]cty.Value{}),
+		crosstests.InferPulumiValue(),
+	)
+}
+
+// Regression test for [pulumi/pulumi-terraform-bridge#1970] and [pulumi/pulumi-terraform-bridge#1964]
+func TestOptionalSetNotSpecified(t *testing.T) {
+	t.Parallel()
+	crosstests.Create(t,
+		map[string]*schema.Schema{
+			"f0": {
+				Optional: true,
+				Type:     schema.TypeSet,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"x": {Optional: true, Type: schema.TypeString},
+					},
+				},
+			},
+		},
+		cty.ObjectVal(map[string]cty.Value{}),
+		crosstests.InferPulumiValue(),
+	)
+}
+
+// Regression test for [pulumi/pulumi-terraform-bridge#1915]
+func TestInputsEqualEmptyList(t *testing.T) {
+	t.Parallel()
+	for _, maxItems := range []int{0, 1} {
+		t.Run(fmt.Sprintf("MaxItems: %v", maxItems), func(t *testing.T) {
+			crosstests.Create(t,
+				map[string]*schema.Schema{
+					"f0": {
+						Optional: true,
+						Type:     schema.TypeList,
+						MaxItems: maxItems,
+						Elem: &schema.Resource{
+							Schema: map[string]*schema.Schema{
+								"x": {Optional: true, Type: schema.TypeString},
+							},
+						},
+					},
+				},
+				cty.ObjectVal(map[string]cty.Value{
+					"f0": cty.ListValEmpty(cty.String),
+				}),
+				crosstests.InferPulumiValue(),
+			)
+		})
+	}
+}
+
+// TestAccCloudWatch failed with PlanResourceChange to do a simple Create preview because the state upgrade was
+// unexpectedly called with nil state. Emulate this here to test it does not fail.
+func TestCreateDoesNotPanicWithStateUpgraders(t *testing.T) {
+	t.Parallel()
+	resourceRuleV0 := func() *schema.Resource {
+		return &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"event_bus_name": {
+					Type:     schema.TypeString,
+					Optional: true,
+				},
+				"is_enabled": {
+					Type:     schema.TypeBool,
+					Optional: true,
+				},
+			},
+		}
+	}
+
+	resourceRuleUpgradeV0 := func(ctx context.Context, rawState map[string]any, meta any) (map[string]any, error) {
+		if rawState == nil {
+			rawState = map[string]any{}
+		}
+
+		if rawState["is_enabled"].(bool) { // used to panic here
+			t.Logf("enabled")
+		} else {
+			t.Logf("disabled")
+		}
+
+		return rawState, nil
+	}
+
+	crosstests.Create(t,
+		resourceRuleV0().Schema,
+		cty.ObjectVal(map[string]cty.Value{
+			"event_bus_name": cty.StringVal("default"),
+		}),
+		crosstests.InferPulumiValue(),
+		crosstests.CreateStateUpgrader(1, []schema.StateUpgrader{
+			{
+				Type:    resourceRuleV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: resourceRuleUpgradeV0,
+				Version: 0,
+			},
+		}),
+	)
+}
+
+func TestTimeouts(t *testing.T) {
```

Comment:
This is interesting but this only calls create timeouts. Are we sure that asserts are strong enough in `crosstests` to check that timeouts are propagated similarly? I recall there was quite a complicated story around timeouts. Perhaps out of scope here, but we can x-ref some issues and comment on this.

Reply:
I'm not 100% sure. I'll add an explicit timeout validation block in `assertResourceDataEqual` to verify.

Reply:
Could also be just something to backlog into the quality epic.